### ISSUE TITLE
[QUESTION] - implement Cmd::GoTo(Position::at()) for table

### ISSUE THREAD

**Issue by tramhao** [veeso/tui-realm#26]
My playlist has a component of table, I want to implement Cmd::GoTo(Position::at()) for the table, so I did this:
```rust
    fn perform(&mut self, cmd: Cmd) -> CmdResult {
        match cmd {
            Cmd::GoTo(Position::At(index)) => {
                let prev = self.component.states.list_index;
                self.component.states.list_index = index;
                if prev == self.component.states.list_index {
                    CmdResult::None
                } else {
                    CmdResult::Changed(self.state())
                }
            }
            _ => self.component.perform(cmd),
        }
    }

```
in mockcomponent trait. However, how can I call it from model or app? It's not a attr. Thanks so much!

**Comment by tramhao** [comment]
basically what I want to do is, I want the table to select certain index based on search result from other dialogue. I can find the value and index, but table doesn't implement Cmd::Goto(Position::at()) .

**Comment by veeso** [comment]
It's in my plans to implement the GoTo(Position::at()) for mock components, but I'm totally busy on other projects at the moment.

At the moment you can access component states from the `on()` method and edit the states with `self.states.list_index`, in this case, when matching an incoming event from the application.

**Comment by tramhao** [comment]
so should i register a port for the event?

**Comment by veeso** [comment]
Sorry, I misunderstood the issue I guess.

There's actually no setter for table/list index in stdlib. I'll make a patch tomorrow.

**Comment by tramhao** [comment]
Actually I want to control the table index from model or update. In model or update, I can get the index, then I need to send this as a userevent to component table. 
Sorry that I'm still confusing about the whole workflow.

**Comment by tramhao** [comment]
In here I don't have access to data in my model. What should I do? Thanks so much.

```rust
impl Poll<UserEvent> for Playlist {
    fn poll(&mut self) -> ListenerResult<Option<Event<UserEvent>>> {
        // ... do something ...
        Ok(Some(Event::User(UserEvent::GotoTableIndex(index))))
    }
}


```

**Comment by tramhao** [comment]
In Component trait I can do following:
```rust
            Event::User(UserEvent::GotoTableIndex(index)) => {
                self.perform(Cmd::GoTo(Position::At(index)))
            }

```

But how to send this event from Model?

**Comment by tramhao** [comment]
Good news is I solved this problem. I implement this like the initial_node in treeview. 
Bad news is, I still don't know the flow of event,msg and cmd.

**Comment by veeso** [comment]
I've released tui-realm-stdlib 1.1.5, which adds the `Value` attribute for list and table in order to set the selected item.

I'm concerned that tui-realm is not super-easy to understand. It is based on the Elm architecture, which took me a while to be understood.
Actually, it's not really different from tui-realm 0.x though. I just decouple a few things, such as `Cmd` and `Event` in order to make component libraries independent from end-user implementation. In the end basically the `Ports` produce `Event`, which are fetched by the `Application` when you call `Tick`. Events are forwarded to subscribed component (plus the current active component). The `on()` method of `Component` transforms the `Event` into a `Cmd` for the inner `MockComponent`, which then produces a `CmdResult` which must be transformed into a `Msg`.
The `Msg` is finally consumed by the "Model", which, based on the msg kind, can update the model or the view.

**Comment by tramhao** [comment]
Yes, I read the workflow in documents several times. I just don't know, how to implement ports for Model, as Model has access for all the data and functions.

**Comment by veeso** [comment]
Ports should never have access to the model. Ports are for fetching external sources in order to produce events to be consumed by the view. An example is the input listener, but could also be an HTTP client, which must fetch a certain source with a certain interval to produce events for a component.

Actually, ports, except for the input listener, are very unlikely to be used.

**Comment by tramhao** [comment]
> I've released tui-realm-stdlib 1.1.5, which adds the `Value` attribute for list and table in order to set the selected item.
> 
> I'm concerned that tui-realm is not super-easy to understand. It is based on the Elm architecture, which took me a while to be understood. Actually, it's not really different from tui-realm 0.x though. I just decouple a few things, such as `Cmd` and `Event` in order to make component libraries independent from end-user implementation. In the end basically the `Ports` produce `Event`, which are fetched by the `Application` when you call `Tick`. Events are forwarded to subscribed component (plus the current active component). The `on()` method of `Component` transforms the `Event` into a `Cmd` for the inner `MockComponent`, which then produces a `CmdResult` which must be transformed into a `Msg`. The `Msg` is finally consumed by the "Model", which, based on the msg kind, can update the model or the view.

This version works great!
